lib/burnside/intrinsics.cc

```diff
@@ -329,7 +334,20 @@ class FunctionDistance {
     case Conversion::Extend: conversions[extendingResult]++; break;
     }
   }
-
+  // Floating point can be mlir::FloatType or fir::real
+  static unsigned getFloatingPointWidth(mlir::Type t) {
+    if (auto f{t.dyn_cast<mlir::FloatType>()}) return f.getWidth();
+    // FIXME: Get width another way for fir.real/complex
+    // - use fir/KindMapping.h and LLVM::Type
+    // - or use evaluate/type.h
+    if (auto r{t.dyn_cast<fir::RealType>()}) return r.getFKind() * 4;
+    if (auto cplx{t.dyn_cast<fir::CplxType>()}) return cplx.getFKind() * 4;
+    assert(false && "not a floating-point type");
+    return 0;
+  }
+  static bool isFloatingPointType(mlir::Type t) {
+    return t.isa<mlir::FloatType>() || t.isa<fir::RealType>();
+  }
```

Comment:
There should already be something like this that gets built from the .td file ... but, I can't recall what it is called off the top of my head.  Bitwidths and subtype testing, I mean.

Reply:
Mmmh, you mean something generated from the the `AnyRealLike` constraint ?
I looked into generated `tools/flang/include/fir/FIROps.cpp.inc` and `tools/flang/include/fir/FIROps.h.inc` but could not find anything like that.

Looking at `AddfOp::verify` in `FIROps.cpp.inc` (currently line 284), it looks like the `AnyRealLike` part has been directly inlined into `((((v->getType().isa<FloatType>())) || (((v->getType().isa<VectorType>())) && ((v->getType().cast<ShapedType>().getElementType().isa<FloatType>()))) || (((v->getType().isa<TensorType>())) && ((v->getType().cast<ShapedType>().getElementType().isa<FloatType>())))) || ((v->getType().isa<fir::RealType>())))` (yiikes, more complex than my test as it seems to account for vector/tensor types when I only care about scalar).

Reply:
Fair enough. Thanks for digging around.